### ISSUE TITLE
consider running dnsmasq on nodes instead of kube-dns

### ISSUE THREAD

**Issue by bprashanth** [kubernetes/kubernetes#32749]
Bind dnsmasq on every node to cbr0 ip and point the containers resolve.conf to cbr0 instead of kube-dns. Setup a dnsmasq.conf like:

```
#listen on container interface
listen-address=0.0.0.0
interface=eth0
interface=cbr0
user=root

#only use these namesservers
no-resolv
server=/svc.cluster.local/10.0.0.10
server=8.8.4.4
server=8.8.8.8

#static entries
address=/foo.bar/10.0.0.3

log-queries 
log-facility=/tmp/dnsmasq.log
```

At least 2 reasons to consider this: 
- External lookups won't fail if kube-dns/endpoint controller/kube-proxy are down
- Internal lookups will fail fast. From: http://www.thekelleys.org.uk/dnsmasq/docs/dnsmasq-man.html, by default, dnsmasq will send queries to any of the upstream servers it knows about and tries to favour servers that are known to be up. 

As an added bonus, we can expose dnsmasq caching/ttl options through a configmap lookups get served from the same node. 
### Eg

Normal svc lookup with dnsmasq

```
root@dnsutils:/# nslookup echoheadersx.default
Server:     10.180.1.1
Address:    10.180.1.1#53

Name:   echoheadersx.default.svc.cluster.local
Address: 10.0.14.169
```

logs

```
Sep 14 23:06:49 dnsmasq[278]: query[A] echoheadersx.default.default.svc.cluster.local from 10.180.1.225
Sep 14 23:06:49 dnsmasq[278]: forwarded echoheadersx.default.default.svc.cluster.local to 10.0.0.10
Sep 14 23:06:49 dnsmasq[278]: reply echoheadersx.default.default.svc.cluster.local is NXDOMAIN
Sep 14 23:06:49 dnsmasq[278]: query[A] echoheadersx.default.svc.cluster.local from 10.180.1.225
Sep 14 23:06:49 dnsmasq[278]: forwarded echoheadersx.default.svc.cluster.local to 10.0.0.10
Sep 14 23:06:49 dnsmasq[278]: reply echoheadersx.default.svc.cluster.local is 10.0.14.169
```

kill kube-dns

```
$ kubectl scale kube-dns-v19 --replicas=0
```

redo the lookup, it returns in a second

```
root@dnsutils:/# nslookup echoheadersx.default
Server:     10.180.1.1
Address:    10.180.1.1#53

** server can't find echoheadersx.default: NXDOMAIN
```

logs

```
Sep 14 23:07:57 dnsmasq[290]: query[A] echoheadersx.default.default.svc.cluster.local from 10.180.1.225
Sep 14 23:07:57 dnsmasq[290]: query[A] echoheadersx.default.svc.cluster.local from 10.180.1.225
Sep 14 23:07:57 dnsmasq[290]: query[A] echoheadersx.default.cluster.local from 10.180.1.225
Sep 14 23:07:57 dnsmasq[290]: forwarded echoheadersx.default.cluster.local to 8.8.8.8
Sep 14 23:07:57 dnsmasq[290]: forwarded echoheadersx.default.cluster.local to 8.8.4.4
Sep 14 23:07:57 dnsmasq[290]: reply echoheadersx.default.cluster.local is NXDOMAIN
Sep 14 23:07:57 dnsmasq[290]: query[A] echoheadersx.default.google.internal from 10.180.1.225
Sep 14 23:07:57 dnsmasq[290]: forwarded echoheadersx.default.google.internal to 8.8.8.8
Sep 14 23:07:57 dnsmasq[290]: reply echoheadersx.default.google.internal is NXDOMAIN
Sep 14 23:07:57 dnsmasq[290]: query[A] echoheadersx.default.c.kubernetesdev.internal from 10.180.1.225
Sep 14 23:07:57 dnsmasq[290]: forwarded echoheadersx.default.c.kubernetesdev.internal to 8.8.8.8
Sep 14 23:07:57 dnsmasq[290]: reply echoheadersx.default.c.kubernetesdev.internal is NXDOMAIN
Sep 14 23:07:57 dnsmasq[290]: query[A] echoheadersx.default from 10.180.1.225
Sep 14 23:07:57 dnsmasq[290]: forwarded echoheadersx.default to 8.8.8.8
Sep 14 23:07:57 dnsmasq[290]: reply echoheadersx.default is NXDOMAIN
```

But on another host the same lookup takes 15s to fail: 

```
$ root@dnsutils2:/# dig +showsearch echoheaders.default

; <<>> DiG 9.8.1-P1 <<>> +showsearch google.com
;; global options: +cmd
;; connection timed out; no servers could be reached
```

A tcpdump shows that we retry thrice with a 5s timeout: 

```
root@e2e-test-beeps-minion-group-iwp2:~# tcpdump -i cbr0 | grep -i google
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on cbr0, link-type EN10MB (Ethernet), capture size 262144 bytes
22:18:56.278326 IP 10.180.0.134.56072 > 10.0.0.10.domain: 713+ A? google.com.default.svc.cluster.local. (54)
22:19:01.278432 IP 10.180.0.134.56072 > 10.0.0.10.domain: 713+ A? google.com.default.svc.cluster.local. (54)
22:19:06.278565 IP 10.180.0.134.56072 > 10.0.0.10.domain: 713+ A? google.com.default.svc.cluster.local. (54)
```

@kubernetes/sig-network

**Comment by thockin** [comment]
Dude, you stole my idea. :)  I sent myself a TODO just a few hours ago to
investigate this, but I want to go a step further and avoid hitting
conntrack for DNS at all.  We could either expand DNS service IP to
endpoints and HUP dnsmasq or do socket-reuse (with a pool so we get to all
DNS backends - maybe 3x the number of DNS replicas to limit conntrack).

We don't want kubelet to interact with net plugins, so maybe just using the
node IP or a link-local address...

**Comment by eparis** [comment]
@smarterclayton @sdodson I know we already do this in OpenShift. Are there things/traps we should be careful of?

**Comment by sdodson** [comment]
> @smarterclayton @sdodson I know we already do this in OpenShift. Are there things/traps we should be careful of?

We use dnsmasq on each node to send `cluster.local` to the kube service ip where skydns is running. Prior to the kube service endpoint lease management we'd see the default query limit of 150 hit and queries back up when an api endpoint was down. We're setting `strict-order` which may be short circuiting the fail-fast behavior @bprashanth is describing, though I doubt it given that we only have one server defined for `cluster.local`.

In OpenShift this is done via the installer so we set dnsIP in config to point at the ip address of the interface that owns the default route. I'm sure if the kubelet were managing all of this it could be more intelligently to handle ip address changes. We're also setting `domain-needed` though I'm not certain that's necessary.

So in short, everything here sounds good to me.

**Comment by smarterclayton** [comment]
I want special DNS lookups on every node that allows dnsmasq to be bound (as proposed).  We also run DNS as part of Kube-proxy so i want enough flexibility that I can point that cbr0 to the proxy if dnsmasq isn't installed.

**Comment by bprashanth** [comment]
>We use dnsmasq on each node to send cluster.local to the kube service ip where skydns is running. 

Yeah that's why Tim suggested "avoiding conntrack". dnsmasq doesn't support hot reload on dnsconf right? any ideas on how we can get it to go straight to endpoints?

**Comment by LEW21** [comment]
This would be useful for accessing docker registries running inside the Kubernetes cluster from the Kubernetes itself. Currently it's commonly done with a localhost proxy, which is slower and not as elegant as simply using "docker-registry.kube-system".

(BTW, in case somebody looking how to make it work today finds this comment: if you're on AWS, use Route53 private hosted zone ".kube-system")

**Comment by smarterclayton** [comment]
We have that use case as well - resolving service addresses from docker is very valuable.

**Comment by bprashanth** [comment]
Fyi I think @bowei is thinking through the details. I have a WIP that deploys dnsmasq as a static pod and introduces a new dnspolicy + bridge in the 169.254 range for it (https://github.com/kubernetes/kubernetes/pull/34523)

**Comment by luxas** [comment]
Status update on this? Relevant anymore?
Would be cool to have...

**Comment by bowei** [comment]
IMHO, the main issue that needs to be worked out is the address on which the dnsmasq will bind to locally in a way that is accessible to all pods on the node. 

There are several choices: 
* bind to node:53
* bind to a local only VIP (the proposed local service feature) 
* bind to *:53

This looks to be a great fit for the local service feature -- but perhaps should not wait for that to land...

**Comment by bgrant0607** [comment]
Ref #7470, #26707, #45363

**Comment by HcgRandon** [comment]
This absolutely is a requirement. Due to kube-dns not running on each node any cluster that simply has a node in eu and one in us and has kube-dns on one or the other will be completely unusable. The one without the dns server adds massive latency for every single request made. Something as simply as looking up google.com takes seconds in such a setup.

**Comment by HcgRandon** [comment]
Also this would fix this issue https://github.com/kubernetes/kubeadm/issues/697.
And also we would no longer have an issues with using slow antiAffinity preventing reverts like this https://github.com/kubernetes/kubernetes/issues/54164#issuecomment-363153968

**Comment by HcgRandon** [comment]
Without dns on every node I am unable to run my application, or any application that communicates with out of cluster resources in k8s. This is absolutely essential for me.

**Comment by pierreozoux** [comment]
@HcgRandon I really think that kubernetes is made to be deployed on one region, and if you want mutli region, you should have 2 clusters. I mean, it is the way it was designed, of course, you can bend the rules, but you'll hit other problems.

**Comment by HcgRandon** [comment]
@pierreozoux k8s advertises itself as planet scale. And if they don't have a solution to this issue I don't think they can continue saying that. At least in my opinion

**Comment by krmayankk** [comment]
Who would this be achieved using CoreDNS (https://github.com/kubernetes/community/pull/1956) ? I believe i hear CoreDNS is one binary, how would we achieve the equivalent of running dnsmasq per node in CoreDNS ?

**Comment by bowei** [comment]
@HcgRandon the usual answer to multi-regional deployments is to have multiple clusters (see https://github.com/kubernetes/community/tree/master/sig-multicluster).

**Comment by bowei** [comment]
@krmayankk the difficulty is a nice portable way (i.e. within the Kubernetes API) to express "contact the service backend closest to me (e.g. localhost)"

**Comment by krmayankk** [comment]
@bowei my question is like in kube-dns i can run dns-masq per node to avoid always going to kube-dns, can i continue to do so in CoreDns, any blockers with that ?

**Comment by tpounds** [comment]
Cross linking https://github.com/kubernetes/kubernetes/issues/45363#issuecomment-389561488 here which provides details on how we use `dnsmasq` as stage1 container in favor of `pause` to resolve a similar issue.

**Comment by zoidbergwill** [comment]
@krmayankk what would the benefits be over running coredns per node, since it seems like it replaces both the dns-masq and kube-dns containers?

**Comment by stp-ip** [comment]
@krmayankk 
Running a DNS cache such as dns-masq per node is still possible with the switch to CoreDNS. It's still answering DNS queries with the usual methods aka UDP/TCP on port 53. 
The difference is just the implementation. Moving the cache into the DNS server and having the implementation much more maintainable and ready for further iterations.

@zoidbergwill
On the CoreDNS per node side it could simply replace a DNS cache such as dns-masq on each node.
In the long run there is a suggestion to make CoreDNS on the node a default. This would enable to iterate on more efficient DNS transport protocols etc.
This could reduce latency impacts by having UDP queries and DNS queries in general only locally on each host. The actual communication with cluster global CoreDNS would then be open to iteration as both client and server are controlled.
Avenues to iterate would be GRPC based queries (no UDP, encrypted, efficient serialization etc.) and a watch API. So instead of the global cache or the per node cach, each node could watch for the specific records they need optimizing DNS latency even more.

This is a rough suggestion and not a clear roadmap. Efforts are made to clear up the feasibility. This might then turn into a proposal including benchmarks etc.

**Comment by bowei** [comment]
/lifecycle frozen
@prameshj can we link to a more current issue and close this one?

**Comment by stp-ip** [comment]
I don't think there is a more recent one yet, but it might have slipped through my notifications.

@jakexks and me might take a stab at benchmarking the various ideas around DNS setups including running a daemonset as cache. Most likely using Coredns to be able to control both client and server and therefore the protocol etc.

**Comment by blakebarnett** [comment]
I haven't had a chance to clean this up and make it configurable, but here's a patch that lets a DaemonSet run with node-local resolution:

```
diff --git a/nodeup/pkg/model/kubelet.go b/nodeup/pkg/model/kubelet.go
index ef8867fac..d7ddc1d4f 100644
--- a/nodeup/pkg/model/kubelet.go
+++ b/nodeup/pkg/model/kubelet.go
@@ -22,6 +22,7 @@ import (
        "fmt"
        "path"
        "path/filepath"
+       "strings"
        "time"

        "github.com/aws/aws-sdk-go/aws/ec2metadata"
@@ -170,6 +171,26 @@ func (b *KubeletBuilder) buildSystemdEnvironmentFile(kubeletConfig *kops.Kubelet
                flags += " --cloud-config=" + CloudConfigFilePath
        }

+       sess := session.Must(session.NewSession())
+       metadata := ec2metadata.New(sess)
+       localIpv4, err := metadata.GetMetadata("local-ipv4")
+       if err != nil {
+               return nil, fmt.Errorf("error fetching the local-ipv4 address from the ec2 meta-data: %v", err)
+       }
+
+       old_cluster_dns_key := -1
+       flags_array := strings.Fields(flags)
+       for i := range flags_array {
+               if strings.HasPrefix(flags_array[i], "--cluster-dns=") {
+                       old_cluster_dns_key = i
+               }
+       }
+       if old_cluster_dns_key >= 0 {
+               flags_array = append(flags_array[:old_cluster_dns_key], flags_array[old_cluster_dns_key+1:]...)
+               flags = strings.Join(flags_array, " ")
+       }
+       flags += " --cluster-dns=" + localIpv4
+
        if b.UsesCNI() {
                flags += " --cni-bin-dir=" + b.CNIBinDir()
                flags += " --cni-conf-dir=" + b.CNIConfDir()
diff --git a/pkg/model/components/kubelet.go b/pkg/model/components/kubelet.go
index 3d263e97b..6a8549a23 100644
--- a/pkg/model/components/kubelet.go
+++ b/pkg/model/components/kubelet.go
@@ -64,14 +64,6 @@ func (b *KubeletOptionsBuilder) BuildOptions(o interface{}) error {
        clusterSpec.Kubelet.ClusterDomain = clusterSpec.ClusterDNSDomain
        clusterSpec.Kubelet.NonMasqueradeCIDR = clusterSpec.NonMasqueradeCIDR

-       if clusterSpec.Kubelet.ClusterDNS == "" {
-               ip, err := WellKnownServiceIP(clusterSpec, 10)
-               if err != nil {
-                       return err
-               }
-               clusterSpec.Kubelet.ClusterDNS = ip.String()
-       }
-
        if b.Context.IsKubernetesLT("1.7") {
                // babysit-daemons removed in 1.7
                clusterSpec.Kubelet.BabysitDaemons = fi.Bool(true)
```

**Comment by prameshj** [comment]
#45363 seems to be a newer issue.

**Comment by blakebarnett** [comment]
oops, thought this was the related kops issue, deleted my patch comment.

**Comment by bowei** [comment]
Let's merge the discussion on to issue #45363

/close
